File: Tff.FSharp.CodeStock/Tff.FrequentFlyerCalculator.FS/CalculationEngine.fs
namespace Tff.FrequentFlyerCalculator.FS

open System

type Flight = {id: int; startCity: string; endCity: string; flightMiles: int; internationalFlight: bool}
type Customer = {id: int; name: string; isSilver: bool; isGold: bool; internationalTraveler: bool; joinDate: DateTime}

type CalculationEngine(flight: Flight, customer: Customer) = 
    member this.CalculateFrequentFlyerMiles = 
        let isLongStandingCustomer = 
            if customer.joinDate.Year - DateTime.Now.Year > 1 then true
            else false
        match isLongStandingCustomer, customer.isGold, customer.isSilver, customer.internationalTraveler, flight.internationalFlight, customer.name with
            | true,true,false,true,true,_ -> float flight.flightMiles * 1.5
            | true,false,true,true,true,_ -> float flight.flightMiles * 1.25
            | false,true,false,true,true,_ -> float flight.flightMiles * 1.05
            | false,false,true,true,true,_ -> float flight.flightMiles * 1.01
            | _,_,_,_,_,"The Boss" -> float flight.flightMiles * 1.3
            //| _,_,_,_,_,_ -> float flight.flightMiles 

File: Tff.FSharp.CodeStock/Tff.FrequentFlyerCalculator.CS/CalculationEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tff.FrequentFlyerCalculator.CS
{
    public class CalculationEngine
    {
        Flight _flight = null;
        Customer _customer = null;
        public CalculationEngine(Flight flight, Customer customer)
        {
            if (flight == null)
            {
                throw new ArgumentNullException("flight");
            }
            if (customer == null)
            {
                throw new ArgumentNullException("customer");
            }
            _flight = flight;
            _customer = customer;
            
        }

        public double CalculateFrequentFlyerMiles()
        {
            Int32 baseMiles = _flight.FlightMiles;

            if (_customer.JoinDate > DateTime.Now.AddYears(-1))
            {
                if (_customer.IsGold && _customer.InternationalTraveler && _flight.InternationalFlight)
                {
                    return baseMiles * 1.5;
                }
                else if (_customer.IsSilver && _customer.InternationalTraveler && _flight.InternationalFlight)
                {
                    return baseMiles * 1.25;
                }
            }
            else
            {
                if (_customer.IsGold && _customer.InternationalTraveler && _flight.InternationalFlight)
                {
                    return baseMiles * 1.05;

                }
                else if (_customer.IsSilver && _customer.InternationalTraveler && _flight.InternationalFlight)
                {
                    return baseMiles * 1.01;

                }
            }

            if (_customer.Name == "The Boss")
            {
                return baseMiles * 1.3;
            }

            return _flight.FlightMiles;
        }


    }
}
